{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "463           GCA_000176535.2_ASM17653v2\n",
       "464           GCA_000176675.2_ASM17667v2\n",
       "465           GCA_000176695.2_ASM17669v2\n",
       "466           GCA_000190855.1_ASM19085v1\n",
       "467           GCA_000190895.1_ASM19089v1\n",
       "468           GCA_000190915.1_ASM19091v1\n",
       "469           GCA_000191015.1_ASM19101v1\n",
       "470           GCA_000210475.1_ASM21047v1\n",
       "471    GCA_000241975.1_Esch_coli_H397_V1\n",
       "472    GCA_000242035.1_Esch_coli_H494_V1\n",
       "473    GCA_000261405.1_Esch_coli_H730_V1\n",
       "474          GCA_002109645.1_ASM210964v1\n",
       "475          GCA_002109685.1_ASM210968v1\n",
       "476          GCA_002109765.1_ASM210976v1\n",
       "477          GCA_002109775.1_ASM210977v1\n",
       "478          GCA_002109805.1_ASM210980v1\n",
       "479          GCA_002110115.1_ASM211011v1\n",
       "480          GCA_002110165.1_ASM211016v1\n",
       "481          GCA_002110225.1_ASM211022v1\n",
       "482          GCA_002110245.1_ASM211024v1\n",
       "483          GCA_002110265.1_ASM211026v1\n",
       "484          GCA_002110275.1_ASM211027v1\n",
       "485          GCA_002110545.1_ASM211054v1\n",
       "486          GCA_002110555.1_ASM211055v1\n",
       "Name: genome_name, dtype: object"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df = pd.read_csv('assembly_summary.csv', sep='\\t', index_col='assembly_accession')\n",
    "\n",
    "# Get genome name\n",
    "df['genome_name'] = pd.DataFrame(df['ftp_path'].str.split('/', expand=True).iloc[:,-1])\n",
    "\n",
    "# Add O serotype info\n",
    "serotype_re = '[\\:\\040|]([O]\\d{1,3})'\n",
    "df['given_O'] = df['organism_name'].str.extract(serotype_re, expand=False)\n",
    "\n",
    "# Add H serotype info\n",
    "serotype_re2 = '[\\:\\040|]([H]\\d{1,3})'\n",
    "df['given_H'] = df['organism_name'].str.extract(serotype_re2, expand=False)\n",
    "\n",
    "# Remove rows without O/H serotype\n",
    "df = df[df['given_O'].notnull() | df['given_H'].notnull()].copy()\n",
    "\n",
    "temp_df = df.copy()\n",
    "# Filter out incorrect H serotypes\n",
    "df['given_H'].str[1:]\n",
    "\n",
    "filter_s = (df['given_H'].str[1:].astype(float)<=56) | df['given_H'].isnull()\n",
    "filter_s2 = (df['given_O'].str[1:].astype(float)<=186) | df['given_O'].isnull()\n",
    "df = df[filter_s & filter_s2].copy()\n",
    "'''\n",
    "# Remove once that are potentially strain name (might filter out legit genome, disabled for now)\n",
    "df = df[((df.organism_name.str.split('Escherichia coli ').str[-1]) != \\\n",
    "                      (df.infraspecific_name.str.split('=').str[-1]))].copy()\n",
    "'''\n",
    "mapping_df = df\n",
    "temp_df.shape\n",
    "df = mapping_df.merge(temp_df, on='genome_name', how='outer', indicator=True)\n",
    "display(df[df._merge!='both']['genome_name'])\n",
    "# temp_df[temp_df.genome_name.str.contains('001874485')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def summarize_result(result_df):\n",
    "    important_cols = [\n",
    "        'organism_name', 'genome_name', 'infraspecific_name',\n",
    "        'given_O', 'predicted_O', 'O_info',\n",
    "        'given_H', 'predicted_H', 'H_info'\n",
    "    ]\n",
    "    \n",
    "    df = result_df\n",
    "    print('all predictions(%d):' %result_df.shape[0])\n",
    "    '''\n",
    "    no prediction if:\n",
    "        both predicted antigen are None\n",
    "    '''\n",
    "    s1 = (df['predicted_O'].isnull() & df['predicted_H'].isnull())\n",
    "    no_df = df[s1][important_cols]\n",
    "    print('%d no predictions' %no_df.shape[0])\n",
    "    '''\n",
    "    incorrect prediction if:\n",
    "        not in 'no_df'\n",
    "        at least one prediction is wrong\n",
    "    '''\n",
    "    s2 = (\n",
    "        ((df['given_O'].notnull() & df['predicted_O'].notnull()) & (df['given_O']!=df['predicted_O'])) |\n",
    "        ((df['given_H'].notnull() & df['predicted_H'].notnull()) & (df['given_H']!=df['predicted_H']))\n",
    "    )\n",
    "    incorrect_df = df[~s1 & s2][important_cols]\n",
    "    print('%d incorrect predictions' %incorrect_df.shape[0])\n",
    "    '''\n",
    "    correct prediction if:\n",
    "        not in 'no_df'\n",
    "        not in 'incorrect_df'\n",
    "        for each given serotype, the correct prediction is made\n",
    "    '''\n",
    "    s3 = (\n",
    "        (df['given_O'].isnull() | (df['given_O']==df['predicted_O'])) &\n",
    "        (df['given_H'].isnull() | (df['given_H']==df['predicted_H']))\n",
    "    )\n",
    "    correct_df = df[~s1 & ~s2 & s3][important_cols]\n",
    "    print('%d correct predictions' %correct_df.shape[0])\n",
    "    '''\n",
    "    semicorrect prediction if:\n",
    "        not in 'no_df'\n",
    "        not in 'incorrect_df'\n",
    "        not in 'correct_df'\n",
    "        one correct prediction is made, no prediction for the other\n",
    "    '''\n",
    "    s4 = (\n",
    "        (df['given_O'].isnull() | (df['given_O']==df['predicted_O'])) |\n",
    "        (df['given_H'].isnull() | (df['given_H']==df['predicted_H']))\n",
    "    )\n",
    "    semicorrect_df = df[~s1 & ~s2 & ~s3 & s4][important_cols]\n",
    "    print('%d semicorrect predictions' %semicorrect_df.shape[0])\n",
    "    '''\n",
    "    remaining prediction\n",
    "    '''\n",
    "    remaining_df = df[~s1 & ~s2 & ~s3 & ~s4]\n",
    "    print('%d remaining predictions' %remaining_df.shape[0])\n",
    "    \n",
    "    # Summary\n",
    "    correct_count = 0\n",
    "    incorrect_count = 0\n",
    "    given_count = 0\n",
    "    # O serotype\n",
    "    # get number of given serotypes\n",
    "    total_s = df['given_O'].notnull()\n",
    "    num_total = df[total_s].shape[0]\n",
    "    given_count += num_total\n",
    "    print(\"number of given_O serotypes is %d\"\n",
    "            %num_total)\n",
    "    # get number of unpredicted serotype\n",
    "    unpredicted_s = total_s & df['predicted_O'].isnull()\n",
    "    num_unpredicted = (df[unpredicted_s].shape[0])\n",
    "    print(\"number of unpredicted_O serotypes is %d or %.2f%%\"\n",
    "            %(num_unpredicted, num_unpredicted/num_total*100))\n",
    "    correct_s = total_s & ~unpredicted_s & (df['given_O']==df['predicted_O'])\n",
    "    num_correct = df[correct_s].shape[0]\n",
    "    correct_count += num_correct\n",
    "    print(\"number of correctly predicted_O serotypes is %d or %.2f%%\"\n",
    "            %(num_correct, num_correct/num_total*100))\n",
    "    incorrect_s = total_s & ~unpredicted_s & ~correct_s\n",
    "    incorrect_o_df = df[incorrect_s][important_cols]\n",
    "    num_incorrect = incorrect_o_df.shape[0]\n",
    "    incorrect_count += num_incorrect\n",
    "    print(\"number of incorrectly predicted_O serotypes is %d or %.2f%%\"\n",
    "            %(num_incorrect, num_incorrect/num_total*100))\n",
    "    \n",
    "    # H serotype\n",
    "    # get number of given serotypes\n",
    "    total_s = df['given_H'].notnull()\n",
    "    num_total = (df[total_s].shape[0])\n",
    "    given_count += num_total\n",
    "    print(\"number of given_H serotypes is %d\"\n",
    "            %num_total)\n",
    "    # get number of unpredicted serotype\n",
    "    unpredicted_s = total_s & df['predicted_H'].isnull()\n",
    "    num_unpredicted = (df[unpredicted_s].shape[0])\n",
    "    print(\"number of unpredicted_H serotypes is %d or %.2f%%\"\n",
    "            %(num_unpredicted, num_unpredicted/num_total*100))\n",
    "    correct_s = total_s & ~unpredicted_s & (df['given_H']==df['predicted_H'])\n",
    "    num_correct = df[correct_s].shape[0]\n",
    "    correct_count += num_correct\n",
    "    print(\"number of correctly predicted_H serotypes is %d or %.2f%%\"\n",
    "            %(num_correct, num_correct/num_total*100))\n",
    "    incorrect_s = total_s & ~unpredicted_s & ~correct_s\n",
    "    incorrect_h_df = df[incorrect_s][important_cols]\n",
    "    num_incorrect = incorrect_h_df.shape[0]\n",
    "    incorrect_count += num_incorrect\n",
    "    print(\"number of incorrectly predicted_H serotypes is %d or %.2f%%\"\n",
    "            %(num_incorrect, num_incorrect/num_total*100))\n",
    "    print(\"Overall concordance=%.2f%%(%d/%d)\" %(correct_count/given_count*100, correct_count,given_count))\n",
    "    print(\"Overall discrepancies=%.2f%%(%d/%d)\" %(incorrect_count/given_count*100, incorrect_count, given_count))\n",
    "\n",
    "    return no_df, incorrect_df, correct_df, semicorrect_df, incorrect_o_df, incorrect_h_df, df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def result_to_summary(df):\n",
    "    h_interested_cols = ['genome_name', 'given_H', 'predicted_H', 'H_info', 'fliC', 'fllA', 'flkA', 'flmA', 'flnA']\n",
    "    o_interested_cols = ['genome_name', 'given_O', 'predicted_O', 'O_info', 'wzx', 'wzy', 'wzm', 'wzt']\n",
    "    # split into two dfs\n",
    "    def rename_antigen_cols(df, antigen):\n",
    "        df = df.rename(columns={\n",
    "            'given_'+antigen: 'given',\n",
    "            'predicted_'+antigen:'predicted',\n",
    "            antigen+'_info':'info'})\n",
    "        df = df[df['given'].notnull() | df['predicted'].notnull()]\n",
    "        df = df[df['given'] != df['predicted']]\n",
    "        return df\n",
    "    mislabelled_genomes = ['GCA_000695155_1', 'GCA_000234315_2', 'GCA_000671295_1', 'GCA_000520035_1', 'GCA_001940375_1', 'GCA_000285655_3', 'GCA_000617365_2', 'GCA_000617425_1', 'GCA_000619345_1', 'GCA_000632575_1', 'GCA_000617045_2']\n",
    "    h_df = rename_antigen_cols(df[h_interested_cols], 'H')\n",
    "    o_df = rename_antigen_cols(df[o_interested_cols], 'O')\n",
    "    df = h_df.append(o_df, ignore_index=True)\n",
    "    interested_cols = ['genome_name', 'given', 'predicted', 'info']\n",
    "    df = df[interested_cols]\n",
    "    df['explanation'] = np.nan\n",
    "    df.loc[(df.genome_name.apply(lambda x: '_'.join(x.replace('.', '_').split('_')[:3])).isin(mislabelled_genomes)) & df.explanation.isnull(), 'explanation'] = 'potentially mislabelled'\n",
    "    df.loc[(df['info']=='No alignment found') & df.explanation.isnull(), 'explanation'] = 'No matching alignment'\n",
    "    df.loc[df.given.isnull() & df.explanation.isnull(), 'explanation'] = 'Insufficient serotype info'\n",
    "    df.loc[(df.genome_name=='GCA_000617165.2_Ec02-3404') & df.explanation.isnull(), 'explanation'] = 'identified as non-ecoli by reference marker alignment'\n",
    "    df.loc[df['info'].isnull() & df.explanation.isnull(), 'explanation'] = 'File system error'\n",
    "    df.loc[df.explanation.isnull(), 'explanation'] = 'Not sure'\n",
    "    df = df.drop(['info'], axis=1)\n",
    "    df.to_csv('genbank_result_summary.csv', na_rep='N/A')\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all predictions(463):\n",
      "2 no predictions\n",
      "9 incorrect predictions\n",
      "415 correct predictions\n",
      "37 semicorrect predictions\n",
      "0 remaining predictions\n",
      "number of given_O serotypes is 460\n",
      "number of unpredicted_O serotypes is 25 or 5.43%\n",
      "number of correctly predicted_O serotypes is 431 or 93.70%\n",
      "number of incorrectly predicted_O serotypes is 4 or 0.87%\n",
      "number of given_H serotypes is 388\n",
      "number of unpredicted_H serotypes is 15 or 3.87%\n",
      "number of correctly predicted_H serotypes is 367 or 94.59%\n",
      "number of incorrectly predicted_H serotypes is 6 or 1.55%\n",
      "Overall concordance=94.10%(798/848)\n",
      "Overall discrepancies=1.18%(10/848)\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv('output/genbank/output.csv')\n",
    "df['genome_name']=df['index'].str.split('_genomic').str[0]\n",
    "df.drop('index', axis=1, inplace=True)\n",
    "\n",
    "df.columns = ['predicted_O', 'O_info', 'predicted_H', 'H_info', 'wzx', 'wzy', 'wzm', 'wzt', 'fliC', 'fllA', 'flkA', 'flmA', 'flnA', 'genome_name']\n",
    "df.loc[df.predicted_O=='-', 'predicted_O'] = np.nan\n",
    "df.loc[df.predicted_H=='-', 'predicted_H'] = np.nan\n",
    "df2 = mapping_df.copy()\n",
    "df = df2.merge(df, on='genome_name', how='left')\n",
    "results = summarize_result(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file</th>\n",
       "      <th>genbank provided</th>\n",
       "      <th>analysis</th>\n",
       "      <th>ectyper predicted</th>\n",
       "      <th>new predicted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>387</th>\n",
       "      <td>GCA_000681435_lcl.fasta</td>\n",
       "      <td>O25</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>\"novel\":H4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>GCA_000749525_lcl.fasta</td>\n",
       "      <td>O91H21</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O181:H49</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>GCA_000749545_lcl.fasta</td>\n",
       "      <td>O91H21</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O91:H21</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>GCA_000749565_lcl.fasta</td>\n",
       "      <td>O91H21</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O91:H21</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>GCA_000749575_lcl.fasta</td>\n",
       "      <td>O91H21</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O91:H21</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>411</th>\n",
       "      <td>GCA_000819645_lcl.fasta</td>\n",
       "      <td>O78</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O78:H19</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>414</th>\n",
       "      <td>GCA_000835045_lcl.fasta</td>\n",
       "      <td>O157</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O157:H19</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>415</th>\n",
       "      <td>GCA_000835055_lcl.fasta</td>\n",
       "      <td>O157:</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O157:H19</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421</th>\n",
       "      <td>GCA_001268885_lcl.fasta</td>\n",
       "      <td>O74</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O66:H45</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>422</th>\n",
       "      <td>GCA_001268925_lcl.fasta</td>\n",
       "      <td>O76</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>-:H34</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428</th>\n",
       "      <td>GCA_001550005_lcl.fasta</td>\n",
       "      <td>O78</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O17:H18</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>442</th>\n",
       "      <td>GCA_001693635_lcl.fasta</td>\n",
       "      <td>O177:H21</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O177:H21</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456</th>\n",
       "      <td>15-520</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O174:H21</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>457</th>\n",
       "      <td>15-579</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O28:H25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>458</th>\n",
       "      <td>16-164</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>not applicable:not applicable</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459</th>\n",
       "      <td>EC20141070</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O153:H25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460</th>\n",
       "      <td>EC20141082</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O178:H19</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461</th>\n",
       "      <td>EC20150519</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O157:H7</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462</th>\n",
       "      <td>EC20151981</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O178:H19</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463</th>\n",
       "      <td>EC20152030</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>not applicable:not applicable</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464</th>\n",
       "      <td>EC20160010</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O26:H32</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>465</th>\n",
       "      <td>EC20160179</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O153:H25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>466</th>\n",
       "      <td>EC20160284</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O178:H16</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>467</th>\n",
       "      <td>EC20160440</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>?:H26</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>468</th>\n",
       "      <td>EC20160772</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O178:H19</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>469</th>\n",
       "      <td>EC20161490</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O2:H29</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>470</th>\n",
       "      <td>EC20161777</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O17:H18</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        file genbank provided  analysis  \\\n",
       "387  GCA_000681435_lcl.fasta              O25  Serotype   \n",
       "406  GCA_000749525_lcl.fasta           O91H21  Serotype   \n",
       "407  GCA_000749545_lcl.fasta           O91H21  Serotype   \n",
       "408  GCA_000749565_lcl.fasta           O91H21  Serotype   \n",
       "409  GCA_000749575_lcl.fasta           O91H21  Serotype   \n",
       "411  GCA_000819645_lcl.fasta              O78  Serotype   \n",
       "414  GCA_000835045_lcl.fasta             O157  Serotype   \n",
       "415  GCA_000835055_lcl.fasta            O157:  Serotype   \n",
       "421  GCA_001268885_lcl.fasta              O74  Serotype   \n",
       "422  GCA_001268925_lcl.fasta              O76  Serotype   \n",
       "428  GCA_001550005_lcl.fasta              O78  Serotype   \n",
       "442  GCA_001693635_lcl.fasta         O177:H21  Serotype   \n",
       "456                   15-520              NaN  Serotype   \n",
       "457                   15-579              NaN  Serotype   \n",
       "458                   16-164              NaN  Serotype   \n",
       "459               EC20141070              NaN  Serotype   \n",
       "460               EC20141082              NaN  Serotype   \n",
       "461               EC20150519              NaN  Serotype   \n",
       "462               EC20151981              NaN  Serotype   \n",
       "463               EC20152030              NaN  Serotype   \n",
       "464               EC20160010              NaN  Serotype   \n",
       "465               EC20160179              NaN  Serotype   \n",
       "466               EC20160284              NaN  Serotype   \n",
       "467               EC20160440              NaN  Serotype   \n",
       "468               EC20160772              NaN  Serotype   \n",
       "469               EC20161490              NaN  Serotype   \n",
       "470               EC20161777              NaN  Serotype   \n",
       "\n",
       "                 ectyper predicted new predicted  \n",
       "387                     \"novel\":H4           NaN  \n",
       "406                       O181:H49           NaN  \n",
       "407                        O91:H21           NaN  \n",
       "408                        O91:H21           NaN  \n",
       "409                        O91:H21           NaN  \n",
       "411                        O78:H19           NaN  \n",
       "414                       O157:H19           NaN  \n",
       "415                       O157:H19           NaN  \n",
       "421                        O66:H45           NaN  \n",
       "422                          -:H34           NaN  \n",
       "428                        O17:H18           NaN  \n",
       "442                       O177:H21           NaN  \n",
       "456                       O174:H21           NaN  \n",
       "457                        O28:H25           NaN  \n",
       "458  not applicable:not applicable           NaN  \n",
       "459                       O153:H25           NaN  \n",
       "460                       O178:H19           NaN  \n",
       "461                        O157:H7           NaN  \n",
       "462                       O178:H19           NaN  \n",
       "463  not applicable:not applicable           NaN  \n",
       "464                        O26:H32           NaN  \n",
       "465                       O153:H25           NaN  \n",
       "466                       O178:H16           NaN  \n",
       "467                          ?:H26           NaN  \n",
       "468                       O178:H19           NaN  \n",
       "469                         O2:H29           NaN  \n",
       "470                        O17:H18           NaN  "
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_excel('in_silico_summary.xlsx')\n",
    "df = df[df.file.notnull()]\n",
    "df2 = results[-1].copy()\n",
    "df2['file'] = df2.genome_name.apply(lambda x: x.split('.')[0]+'_lcl.fasta')\n",
    "df2.fillna('-', inplace=True)\n",
    "df2['new predicted'] = df2.apply(lambda row: ':'.join([row['predicted_O'], row['predicted_H']]), axis=1)\n",
    "df2 = df2[['file', 'new predicted']]\n",
    "merged_df = df.merge(df2, on='file', how='left')\n",
    "merged_df.to_excel('modified_in_silico_summary.xlsx')\n",
    "merged_df[merged_df['new predicted'].isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>wgs_master</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>taxid</th>\n",
       "      <th>species_taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>infraspecific_name</th>\n",
       "      <th>isolate</th>\n",
       "      <th>version_status</th>\n",
       "      <th>...</th>\n",
       "      <th>H_info</th>\n",
       "      <th>wzx</th>\n",
       "      <th>wzy</th>\n",
       "      <th>wzm</th>\n",
       "      <th>wzt</th>\n",
       "      <th>fliC</th>\n",
       "      <th>fllA</th>\n",
       "      <th>flkA</th>\n",
       "      <th>flmA</th>\n",
       "      <th>flnA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows Ã— 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [bioproject, biosample, wgs_master, refseq_category, taxid, species_taxid, organism_name, infraspecific_name, isolate, version_status, assembly_level, release_type, genome_rep, seq_rel_date, asm_name, submitter, gbrs_paired_asm, paired_asm_comp, ftp_path, excluded_from_refseq, relation_to_type_material, genome_name, given_O, given_H, predicted_O, O_info, predicted_H, H_info, wzx, wzy, wzm, wzt, fliC, fllA, flkA, flmA, flnA]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 37 columns]"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = results[-1]\n",
    "df[df.genome_name.str.contains('001693635')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
