{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# General functions\n",
    "def get_diff_row(df1, df2, on):\n",
    "    df = df1.merge(df2, on=on, how='outer', indicator=True)\n",
    "    return (df[df._merge!='both'])\n",
    "\n",
    "def summarize_result(result_df):\n",
    "    \n",
    "    df = result_df.copy()\n",
    "    # Summary\n",
    "    def getSerotypeSummary(antigen, df):\n",
    "        # get number of given serotypes\n",
    "        total_s = df['given_'+antigen].notnull()\n",
    "        num_total = df[total_s].shape[0]\n",
    "        print(\"number of given_\"+antigen+\" serotypes is %d\"\n",
    "                %num_total)\n",
    "        # get number of unpredicted serotype\n",
    "        unpredicted_s = total_s & df['predicted_'+antigen].isnull()\n",
    "        num_unpredicted = (df[unpredicted_s].shape[0])\n",
    "        print(\"number of unpredicted_\"+antigen+\" serotypes is %d or %.2f%%\"\n",
    "                %(num_unpredicted, num_unpredicted/num_total*100))\n",
    "        correct_s = total_s & ~unpredicted_s & (df['given_'+antigen]==df['predicted_'+antigen])\n",
    "        correct_df = df[correct_s]\n",
    "        num_correct = df[correct_s].shape[0]\n",
    "        print(\"number of correctly predicted_\"+antigen+\" serotypes is %d or %.2f%%\"\n",
    "                %(num_correct, num_correct/num_total*100))\n",
    "        incorrect_s = total_s & ~unpredicted_s & ~correct_s\n",
    "        num_incorrect = df[incorrect_s].shape[0]\n",
    "        print(\"number of incorrectly predicted_\"+antigen+\" serotypes is %d or %.2f%%\"\n",
    "                %(num_incorrect, num_incorrect/num_total*100))\n",
    "        result_dfs = {\n",
    "            'concordance': df[correct_s],\n",
    "            'discrepancies': df[incorrect_s],\n",
    "            'unpredicted': df[unpredicted_s]\n",
    "        }\n",
    "        return result_dfs, num_total, num_correct, num_incorrect\n",
    "    result_dict = {}\n",
    "    given_count = 0\n",
    "    correct_count = 0\n",
    "    incorrect_count = 0\n",
    "    for antigen in ['O', 'H']:\n",
    "        result_dfs, num_total, num_correct, num_incorrect = getSerotypeSummary(antigen, df)\n",
    "        result_dict[antigen] = result_dfs\n",
    "        given_count += num_total; correct_count+=num_correct;incorrect_count+=num_incorrect\n",
    "    print(\"Overall concordance=%.2f%%(%d/%d)\" %(correct_count/given_count*100, correct_count,given_count))\n",
    "    print(\"Overall discrepancies=%.2f%%(%d/%d)\" %(incorrect_count/given_count*100, incorrect_count, given_count))\n",
    "    result_dict['all'] = result_df\n",
    "    return result_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Genbank only functions\n",
    "def parse_assembly_summary(file):\n",
    "    df = pd.read_csv(file, sep='\\t', index_col='assembly_accession')\n",
    "    \n",
    "    # Get genome name\n",
    "    df['genome_name'] = pd.DataFrame(df['ftp_path'].str.split('/', expand=True).iloc[:,-1])\n",
    "\n",
    "    # Add O serotype info\n",
    "    serotype_re = '(?<![A-z])(O\\d{1,3})'\n",
    "    df['given_O'] = df['organism_name'].str.extract(serotype_re, expand=False)\n",
    "    df.given_O.fillna(df['infraspecific_name'].str.extract(serotype_re, expand=False), inplace=True)\n",
    "\n",
    "    # Add H serotype info\n",
    "    serotype_re2 = '\\:(H\\d{1,2})'\n",
    "    df['given_H'] = df['organism_name'].str.extract(serotype_re2, expand=False)\n",
    "    df.given_H.fillna(df['infraspecific_name'].str.extract(serotype_re2, expand=False), inplace=True)\n",
    "\n",
    "    # Remove rows without serotype\n",
    "    df = df[df['given_O'].notnull() | df['given_H'].notnull()].copy()\n",
    "    \n",
    "    # Filter out incorrect H serotypes\n",
    "    filter_s = (df['given_H'].str[1:].astype(float)<=56) | df['given_H'].isnull()\n",
    "    filter_s2 = (df['given_O'].str[1:].astype(float)<=186) | df['given_O'].isnull()\n",
    "    df = df[filter_s & filter_s2]\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Enterobase only functions\n",
    "def remove_blacklist_genomes(df):\n",
    "    start_count = df.shape[0]\n",
    "    print('Start with %d rows' %(start_count))\n",
    "    df = df[~df.blacklisted]\n",
    "    print('%d rows removed for blacklisted genome' %(start_count-df.shape[0]))\n",
    "    print('End with %d rows' %(df.shape[0]))\n",
    "    return df\n",
    "\n",
    "def mark_blacklist_genomes(df, blacklist_file):\n",
    "    blacklist_df = pd.read_csv(blacklist_file)\n",
    "    df = df.merge(blacklist_df, on='genome_name', how='outer', indicator=True)\n",
    "    df['blacklisted'] = df._merge!='left_only'\n",
    "    df.drop('_merge', axis=1, inplace=True)\n",
    "    return df\n",
    "\n",
    "def read_info_df(file):\n",
    "    # Read info df\n",
    "    df = pd.read_csv(file).drop('Unnamed: 0', axis=1)\n",
    "    df.columns = ['genome_name', 'given_H', 'given_O', 'serotype_tag']\n",
    "    df.given_H='H'+df.loc[df.given_H.notnull()].given_H.astype(int).astype(str)\n",
    "    df.given_O='O'+df.loc[df.given_O.notnull()].given_O.astype(int).astype(str)\n",
    "    df = mark_blacklist_genomes(df, 'blacklist.csv')\n",
    "    return df\n",
    "\n",
    "def read_output_file(file):\n",
    "    # Read from result file\n",
    "    df = pd.read_csv(file)\n",
    "    df = df[['index', 'O_prediction', 'O_info', 'H_prediction', 'H_info']]\n",
    "    df.columns = ['genome_name', 'predicted_O', 'O_info', 'predicted_H', 'H_info']\n",
    "    df.loc[df.predicted_O=='-', 'predicted_O'] = np.nan\n",
    "    df.loc[df.predicted_H=='-', 'predicted_H'] = np.nan\n",
    "    return df\n",
    "\n",
    "def create_merge_df(info_file, output_file):\n",
    "    # merge with info file\n",
    "    df = read_output_file(output_file)\n",
    "    df2 = read_info_df(info_file)\n",
    "    df = df.merge(df2, on='genome_name', how='inner')\n",
    "    df = df[['genome_name', 'given_O', 'predicted_O', 'O_info', 'given_H', 'predicted_H', 'H_info', 'serotype_tag', 'blacklisted']]\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genbank 97 97:\n",
      "number of given_O serotypes is 535\n",
      "number of unpredicted_O serotypes is 26 or 4.86%\n",
      "number of correctly predicted_O serotypes is 504 or 94.21%\n",
      "number of incorrectly predicted_O serotypes is 5 or 0.93%\n",
      "number of given_H serotypes is 387\n",
      "number of unpredicted_H serotypes is 15 or 3.88%\n",
      "number of correctly predicted_H serotypes is 368 or 95.09%\n",
      "number of incorrectly predicted_H serotypes is 4 or 1.03%\n",
      "Overall concordance=94.58%(872/922)\n",
      "Overall discrepancies=0.98%(9/922)\n"
     ]
    }
   ],
   "source": [
    "print(\"Genbank 97 97:\")\n",
    "mapping_df = parse_assembly_summary('assembly_summary.csv')[['genome_name', 'organism_name', 'infraspecific_name','given_O', 'given_H']]\n",
    "output_df = pd.read_csv('output/genbank/output.csv')\n",
    "output_df['genome_name']=output_df['index'].str.split('_genomic').str[0]\n",
    "output_df.drop('index', axis=1, inplace=True)\n",
    "output_df = output_df[['O_prediction', 'O_info', 'H_prediction', 'H_info', 'genome_name']]\n",
    "output_df.columns = ['predicted_O', 'O_info', 'predicted_H', 'H_info', 'genome_name']\n",
    "output_df.loc[output_df.predicted_O=='-', 'predicted_O'] = np.nan\n",
    "output_df.loc[output_df.predicted_H=='-', 'predicted_H'] = np.nan\n",
    "merge_df = mapping_df.merge(output_df, on='genome_name', how='inner')\n",
    "results = summarize_result(merge_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genbank 90 50:\n",
      "number of given_O serotypes is 535\n",
      "number of unpredicted_O serotypes is 15 or 2.80%\n",
      "number of correctly predicted_O serotypes is 515 or 96.26%\n",
      "number of incorrectly predicted_O serotypes is 5 or 0.93%\n",
      "number of given_H serotypes is 387\n",
      "number of unpredicted_H serotypes is 2 or 0.52%\n",
      "number of correctly predicted_H serotypes is 381 or 98.45%\n",
      "number of incorrectly predicted_H serotypes is 4 or 1.03%\n",
      "Overall concordance=97.18%(896/922)\n",
      "Overall discrepancies=0.98%(9/922)\n"
     ]
    }
   ],
   "source": [
    "print(\"Genbank 90 50:\")\n",
    "mapping_df = parse_assembly_summary('assembly_summary.csv')[['genome_name', 'organism_name', 'infraspecific_name','given_O', 'given_H']]\n",
    "output_df = pd.read_csv('output/genbank9050/output.csv')\n",
    "output_df['genome_name']=output_df['index'].str.split('_genomic').str[0]\n",
    "output_df.drop('index', axis=1, inplace=True)\n",
    "output_df = output_df[['O_prediction', 'O_info', 'H_prediction', 'H_info', 'genome_name']]\n",
    "output_df.columns = ['predicted_O', 'O_info', 'predicted_H', 'H_info', 'genome_name']\n",
    "output_df.loc[output_df.predicted_O=='-', 'predicted_O'] = np.nan\n",
    "output_df.loc[output_df.predicted_H=='-', 'predicted_H'] = np.nan\n",
    "merge_df = mapping_df.merge(output_df, on='genome_name', how='inner')\n",
    "results = summarize_result(merge_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enterobase 97 97:\n",
      "Start with 5844 rows\n",
      "489 rows removed for blacklisted genome\n",
      "End with 5355 rows\n",
      "number of given_O serotypes is 5223\n",
      "number of unpredicted_O serotypes is 827 or 15.83%\n",
      "number of correctly predicted_O serotypes is 4233 or 81.05%\n",
      "number of incorrectly predicted_O serotypes is 163 or 3.12%\n",
      "number of given_H serotypes is 2570\n",
      "number of unpredicted_H serotypes is 44 or 1.71%\n",
      "number of correctly predicted_H serotypes is 2384 or 92.76%\n",
      "number of incorrectly predicted_H serotypes is 142 or 5.53%\n",
      "Overall concordance=84.91%(6617/7793)\n",
      "Overall discrepancies=3.91%(305/7793)\n"
     ]
    }
   ],
   "source": [
    "print(\"Enterobase 97 97:\")\n",
    "merge_df = create_merge_df('enterobase_serotype.csv', 'output/enterobase/output.csv')\n",
    "merge_df = remove_blacklist_genomes(merge_df)\n",
    "results1 = summarize_result(merge_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enterobase 90 50:\n",
      "Start with 5844 rows\n",
      "489 rows removed for blacklisted genome\n",
      "End with 5355 rows\n",
      "number of given_O serotypes is 5223\n",
      "number of unpredicted_O serotypes is 532 or 10.19%\n",
      "number of correctly predicted_O serotypes is 4474 or 85.66%\n",
      "number of incorrectly predicted_O serotypes is 217 or 4.15%\n",
      "number of given_H serotypes is 2570\n",
      "number of unpredicted_H serotypes is 6 or 0.23%\n",
      "number of correctly predicted_H serotypes is 2416 or 94.01%\n",
      "number of incorrectly predicted_H serotypes is 148 or 5.76%\n",
      "Overall concordance=88.41%(6890/7793)\n",
      "Overall discrepancies=4.68%(365/7793)\n"
     ]
    }
   ],
   "source": [
    "print(\"Enterobase 90 50:\")\n",
    "merge_df = create_merge_df('enterobase_serotype.csv', 'output/enterobase 9050/output.csv')\n",
    "merge_df = remove_blacklist_genomes(merge_df)\n",
    "results2 = summarize_result(merge_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enterobase 90 50:\n",
      "number of given_O serotypes is 5703\n",
      "number of unpredicted_O serotypes is 577 or 10.12%\n",
      "number of correctly predicted_O serotypes is 4615 or 80.92%\n",
      "number of incorrectly predicted_O serotypes is 511 or 8.96%\n",
      "number of given_H serotypes is 2897\n",
      "number of unpredicted_H serotypes is 6 or 0.21%\n",
      "number of correctly predicted_H serotypes is 2493 or 86.05%\n",
      "number of incorrectly predicted_H serotypes is 398 or 13.74%\n",
      "Overall concordance=82.65%(7108/8600)\n",
      "Overall discrepancies=10.57%(909/8600)\n"
     ]
    }
   ],
   "source": [
    "print(\"Enterobase 90 50:\")\n",
    "merge_df = create_merge_df('enterobase_serotype.csv', 'output/enterobase 9050/output.csv')\n",
    "summarize_result(merge_df)['all'].to_csv('enterobase_90_50_with_blacklist.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_name</th>\n",
       "      <th>given_O</th>\n",
       "      <th>predicted_O</th>\n",
       "      <th>O_info</th>\n",
       "      <th>given_H</th>\n",
       "      <th>predicted_H</th>\n",
       "      <th>H_info</th>\n",
       "      <th>serotype_tag</th>\n",
       "      <th>blacklisted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>217</td>\n",
       "      <td>217</td>\n",
       "      <td>217</td>\n",
       "      <td>217</td>\n",
       "      <td>86</td>\n",
       "      <td>217</td>\n",
       "      <td>217</td>\n",
       "      <td>217</td>\n",
       "      <td>217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>217</td>\n",
       "      <td>101</td>\n",
       "      <td>82</td>\n",
       "      <td>2</td>\n",
       "      <td>33</td>\n",
       "      <td>39</td>\n",
       "      <td>1</td>\n",
       "      <td>159</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>ESC_GA6873AA_AS</td>\n",
       "      <td>O157</td>\n",
       "      <td>O8</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>H10</td>\n",
       "      <td>H19</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>O157</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>26</td>\n",
       "      <td>166</td>\n",
       "      <td>8</td>\n",
       "      <td>16</td>\n",
       "      <td>217</td>\n",
       "      <td>20</td>\n",
       "      <td>217</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            genome_name given_O predicted_O           O_info given_H  \\\n",
       "count               217     217         217              217      86   \n",
       "unique              217     101          82                2      33   \n",
       "top     ESC_GA6873AA_AS    O157          O8  Alignment found     H10   \n",
       "freq                  1      21          26              166       8   \n",
       "\n",
       "       predicted_H           H_info serotype_tag blacklisted  \n",
       "count          217              217          217         217  \n",
       "unique          39                1          159           1  \n",
       "top            H19  Alignment found         O157       False  \n",
       "freq            16              217           20         217  "
      ]
     },
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_name</th>\n",
       "      <th>given_O</th>\n",
       "      <th>predicted_O</th>\n",
       "      <th>O_info</th>\n",
       "      <th>given_H</th>\n",
       "      <th>predicted_H</th>\n",
       "      <th>H_info</th>\n",
       "      <th>serotype_tag</th>\n",
       "      <th>blacklisted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1874</th>\n",
       "      <td>ESC_GA8888AA_AS</td>\n",
       "      <td>O162</td>\n",
       "      <td>O89</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>H9</td>\n",
       "      <td>H9</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>O162:H9</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2375</th>\n",
       "      <td>ESC_HA0149AA_AS</td>\n",
       "      <td>O42</td>\n",
       "      <td>O89</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H37</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>O42</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3202</th>\n",
       "      <td>ESC_HA7811AA_AS</td>\n",
       "      <td>O2</td>\n",
       "      <td>O89</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>H25</td>\n",
       "      <td>H9</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>O2:H25</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3230</th>\n",
       "      <td>ESC_HA7913AA_AS</td>\n",
       "      <td>O28</td>\n",
       "      <td>O89</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H9</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>O28</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3281</th>\n",
       "      <td>ESC_HA7965AA_AS</td>\n",
       "      <td>O101</td>\n",
       "      <td>O89</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>H12</td>\n",
       "      <td>H9</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>O101:H12</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3317</th>\n",
       "      <td>ESC_HA8003AA_AS</td>\n",
       "      <td>O9</td>\n",
       "      <td>O89</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>H12</td>\n",
       "      <td>H9</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>O9:H12</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3578</th>\n",
       "      <td>ESC_HA8332AA_AS</td>\n",
       "      <td>O99</td>\n",
       "      <td>O89</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>H16</td>\n",
       "      <td>H9</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>O99:H16</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3645</th>\n",
       "      <td>ESC_HA8399AA_AS</td>\n",
       "      <td>O101</td>\n",
       "      <td>O89</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>H37</td>\n",
       "      <td>H37</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>O101:H37</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3841</th>\n",
       "      <td>ESC_HA8598AA_AS</td>\n",
       "      <td>O14</td>\n",
       "      <td>O89</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H9</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>O14</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3846</th>\n",
       "      <td>ESC_HA8603AA_AS</td>\n",
       "      <td>O101</td>\n",
       "      <td>O89</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H37</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>O101</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3994</th>\n",
       "      <td>ESC_HA8768AA_AS</td>\n",
       "      <td>O135</td>\n",
       "      <td>O89</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H37</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>O135</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4038</th>\n",
       "      <td>ESC_HA8812AA_AS</td>\n",
       "      <td>O116</td>\n",
       "      <td>O89</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>H10</td>\n",
       "      <td>H10</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>O116:H10</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4039</th>\n",
       "      <td>ESC_HA8813AA_AS</td>\n",
       "      <td>O101</td>\n",
       "      <td>O89</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>H10</td>\n",
       "      <td>H10</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>O101:H10</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4065</th>\n",
       "      <td>ESC_HA8990AA_AS</td>\n",
       "      <td>O101</td>\n",
       "      <td>O89</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H10</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>O101</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4715</th>\n",
       "      <td>ESC_IA7098AA_AS</td>\n",
       "      <td>O101</td>\n",
       "      <td>O89</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>H9</td>\n",
       "      <td>H9</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>O101:H9</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4721</th>\n",
       "      <td>ESC_IA7147AA_AS</td>\n",
       "      <td>O101</td>\n",
       "      <td>O89</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>H9</td>\n",
       "      <td>H9</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>O101:H9</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          genome_name given_O predicted_O           O_info given_H  \\\n",
       "1874  ESC_GA8888AA_AS    O162         O89  Alignment found      H9   \n",
       "2375  ESC_HA0149AA_AS     O42         O89  Alignment found     NaN   \n",
       "3202  ESC_HA7811AA_AS      O2         O89  Alignment found     H25   \n",
       "3230  ESC_HA7913AA_AS     O28         O89  Alignment found     NaN   \n",
       "3281  ESC_HA7965AA_AS    O101         O89  Alignment found     H12   \n",
       "3317  ESC_HA8003AA_AS      O9         O89  Alignment found     H12   \n",
       "3578  ESC_HA8332AA_AS     O99         O89  Alignment found     H16   \n",
       "3645  ESC_HA8399AA_AS    O101         O89  Alignment found     H37   \n",
       "3841  ESC_HA8598AA_AS     O14         O89  Alignment found     NaN   \n",
       "3846  ESC_HA8603AA_AS    O101         O89  Alignment found     NaN   \n",
       "3994  ESC_HA8768AA_AS    O135         O89  Alignment found     NaN   \n",
       "4038  ESC_HA8812AA_AS    O116         O89  Alignment found     H10   \n",
       "4039  ESC_HA8813AA_AS    O101         O89  Alignment found     H10   \n",
       "4065  ESC_HA8990AA_AS    O101         O89  Alignment found     NaN   \n",
       "4715  ESC_IA7098AA_AS    O101         O89  Alignment found      H9   \n",
       "4721  ESC_IA7147AA_AS    O101         O89  Alignment found      H9   \n",
       "\n",
       "     predicted_H           H_info serotype_tag  blacklisted  \n",
       "1874          H9  Alignment found      O162:H9        False  \n",
       "2375         H37  Alignment found          O42        False  \n",
       "3202          H9  Alignment found       O2:H25        False  \n",
       "3230          H9  Alignment found          O28        False  \n",
       "3281          H9  Alignment found     O101:H12        False  \n",
       "3317          H9  Alignment found       O9:H12        False  \n",
       "3578          H9  Alignment found      O99:H16        False  \n",
       "3645         H37  Alignment found     O101:H37        False  \n",
       "3841          H9  Alignment found          O14        False  \n",
       "3846         H37  Alignment found         O101        False  \n",
       "3994         H37  Alignment found         O135        False  \n",
       "4038         H10  Alignment found     O116:H10        False  \n",
       "4039         H10  Alignment found     O101:H10        False  \n",
       "4065         H10  Alignment found         O101        False  \n",
       "4715          H9  Alignment found      O101:H9        False  \n",
       "4721          H9  Alignment found      O101:H9        False  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = results1['O']['discrepancies'].merge(results2['O']['discrepancies'], on='genome_name', how='outer', indicator=True)\n",
    "df = df.loc[df._merge!='both']\n",
    "df[['genome_name', 'given_O_x', 'predicted_O_x', 'given_O_y', 'predicted_O_y', '_merge']]\n",
    "df = results2['all']\n",
    "df = results2['O']['discrepancies']\n",
    "display(df.describe())\n",
    "df.loc[df.predicted_O=='O89']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'fasta'"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "my_file = \"/home/sam/Projects/Experiment/Data/genbank_genomes/GCA_000008865.1_ASM886v1_genomic.fna\"  # Obviously not FASTA\n",
    "\n",
    "def get_genome_format(filename):\n",
    "    for fm in ['fastq', 'fasta']:\n",
    "        with open(filename, \"r\") as handle:\n",
    "            data = SeqIO.parse(handle, fm)\n",
    "            if any(data):\n",
    "                return fm\n",
    "    return False\n",
    "\n",
    "get_genome_format(my_file)\n",
    "# False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
